frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -561,9 +576,69 @@ function ReaderHighlight:_resetHoldTimer(clear)
     end
 end
 
+function ReaderHighlight:getRegionalZoomCenter(pos)
+    if not self.dimen then
+        self.dimen = self.ui.dimen
+    end
+    local page_size = self.ui.document:getNativePageDimensions(pos.page)
+    local margin = self.ui.document.configurable.page_margin * Screen:getDPI()
+    local zoom = 2*self.dimen.w / page_size.w
+    zoom = zoom/(1 + 3*margin/zoom/page_size.w)
+    local x_ratio = (pos.x/ page_size.w) -- x center ratio
+    local y_ratio = (pos.y / page_size.h) -- y center ratio
+       
+    return zoom, x_ratio, y_ratio
+    
+end
+
+
 function ReaderHighlight:onHold(arg, ges)
     -- disable hold gesture if highlighting is disabled
-    if self.view.highlight.disabled then return false end
+    if self.view.highlight.disabled then
+        if self.view.highlight.zoom then
+            --zoom the hold area
+            self:clear() -- clear previous highlight (delayed clear may not have done it yet)
+            self.hold_ges_pos = ges.pos -- remember hold original gesture position
+            self.hold_pos = self.view:screenToPageTransform(ges.pos)
+            logger.dbg("hold position in page ", self.hold_pos)
+            if not self.hold_pos then
+                logger.dbg("not inside page area")
+                return false
+            end
+            
+            -- Screen:shot("/home/m/1.png")
+            logger.dbg("hold on image")
+
+            local getImage = require("document/koptinterface")
```

Comment:
this include might break crengine document (epubs) and is unnecessary for mupdf based document as it is already loaded.
you can check for this and use the preloaded one, see:
https://github.com/koreader/koreader/blob/7c1a716d57b092bb786eeec3ca9ecb5405a8195d/frontend/apps/reader/modules/readerconfig.lua#L16-L18

with something like: (untested)
```lua
if self.ui.document.koptinterface ~= nil then
    local Image = self.ui.document.koptinterface:getNativePageImage(self.ui.document, self.hold_pos.page)
```

Reply:
I change it and it worked but what should I do it koptinterface is nill...
should I add:- 
```
else
local koptinterface = require("document/koptinterface")
local image = getImage:getNativePageImage(self.ui.document, self.hold_pos.page)
```


sorry I'm new to the app and learning curve need time to let me know how thinks work..

Reply:
> should I add...

No. The documents that work with kopt have it (pdfdocument, djvudocument) - those who don't do not have it (picdocument). Using it yourself would just crash on picdocument that can't.

Also, pdfdocument & co have some wrapping functions that use it internally:
https://github.com/koreader/koreader/blob/db843929f0893383a3dcddbb6d77708691b8e393/frontend/document/pdfdocument.lua#L124-L126
that overrides generic functions that return nil when not available:
https://github.com/koreader/koreader/blob/db843929f0893383a3dcddbb6d77708691b8e393/frontend/document/document.lua#L286-L288

So, ideally, higher level code like readerhighlight should not need to know the document works with koptinterface or not: ask, if you get nil, it doesn't work.